tests: e2e/integration tests

otron opened the pull request:
* Adds E2E/integration tests in `tests/e2e/`, and necessary changes to get them running locally and on Travis.
* Parametrizes the E2E tests via an environment variable and override of pytest's `pytest_generate_tests` in `tests/e2e/conftest.py`.
* Changes some existing tests to work with `pytest-flask`, which provides useful fixtures for e2e-testing.

jirikuncar commented on setup.py:
```diff
@@ -42,6 +42,7 @@
     'pytest-cov>=1.8.0',
     'pytest-pep8>=1.0.6',
     'pytest>=2.8.0',
+    'pytest-flask>=0.10.',
```
`0.10.0`?

otron replied:
Yup! My bad.

lnielsen commented on tests/e2e/conftest.py:
```diff
@@ -0,0 +1,94 @@
+
+import os
+
+from flask import Flask
```
Please check import sorting

otron replied:
With `Isort`, you mean? 'Cause as it is it's checked by kwalitee :p

lnielsen replied:
Hehe, yes with isort :-)

otron commented:
The actual code is ready to be reviewed. The commits still need to be rebased into, well, I don't know how many would be sensible here.

The Python 3.3/requirements=release build fails b/c the E2E tests time out when run with PhantomJS. Not sure why that happens.

otron commented:
Sometimes an E2E test run will timeout in a job, causing the entire build to fail.
E.g. https://travis-ci.org/inveniosoftware/invenio-accounts/jobs/94359708, which is the only failed job in the https://travis-ci.org/inveniosoftware/invenio-accounts/builds/94359695 build.

jirikuncar commented:
@otron maybe there is a way how py.test can re-run test which raises specific exception.

otron commented:
@jirikuncar it doesn't raise any particular exception though. Travis kills the job after 10 minutes.

jirikuncar commented on tests/e2e/conftest.py:
```diff
@@ -0,0 +1,109 @@
+    # Create instance of webdriver.`request.param`()
+    browser = getattr(webdriver, request.param)()
+    # If the test fails it's be useful to know which browser was used.
+    print("Webdriver browser=", request.param)
```
using `print` is not the best way to to detect fail - you should have this information in traceback context.

otron replied:
You are right. Turns out it's also not required, as `pytest` prints the value of `env_browser` as a part of the failure information:
```
live_server = <LiveServer listening at http://localhost:53179>
env_browser = <selenium.webdriver.firefox.webdriver.WebDriver (session="aed62c3e-8180-0341-b571-eb261a5d987e")>

    def test_user_registration(live_server, env_browser):
        """E2E user registration and login test."""
        browser = env_browser
>       assert False
```

otron commented:
Travis build fails b/c the e2e tests timeout on random jobs :(

jirikuncar commented on tests/test_utils.py:
```diff
@@ -93,16 +95,20 @@ def test_client_authenticated(app):
-    # Can't access the datastore outside of an application context
-    with pytest.raises(RuntimeError):
-        testutils.create_test_user(email, password)
+    # If pytest-flask isn't available we can't access the datastore outside of
```
since we `pytest-flask` is part of test dependencies i'm not sure why do we need to keep it here.

otron replied:
Only reason I can think of is as a helpful reminder to future developers who might not be aware of why this repo's tests are different than the ones in repos not using pytest-flask. But this probably isn't the place to include a tidbit like that.

jirikuncar replied:
Testing section in docs might be a better place.

jirikuncar commented on tests/e2e/conftest.py:
```diff
@@ -0,0 +1,120 @@
+    with app.app_context():
+        if not database_exists(str(db.engine.url)):
+            create_database(str(db.engine.url))
+        db.drop_all()
```
it should not be necessary

otron replied:
Which part?

jirikuncar replied:
`db.drop_all()`

otron replied:
While it _shouldn't_ be necessary, I feel it's one of those "better safe than having tests inexplicably fail and then pass on the next run because of some weirdness left in the database"-situations.

jirikuncar replied:
It means that you haven't cleaned your db on tear down and your test MUST necessarily fail then.

otron replied:
Or the process could have been interrupted for some reason, causing the
teardown stage to be skipped.

On 10 December 2015 at 09:13, Jiri Kuncar <notifications@github.com> wrote:

> In tests/e2e/conftest.py
> <https://github.com/inveniosoftware/invenio-accounts/pull/53#discussion_r47198588>
> :
>
> > +            'SQLALCHEMY_DATABASE_URI', 'sqlite:///test.db'),
> > +        TESTING=True,
> > +        LOGIN_DISABLED=False,
> > +    )
> > +    FlaskCLI(app)
> > +    Menu(app)
> > +    Babel(app)
> > +    Mail(app)
> > +    InvenioDB(app)
> > +    InvenioAccounts(app)
> > +    app.register_blueprint(blueprint)
> > +
> > +    with app.app_context():
> > +        if not database_exists(str(db.engine.url)):
> > +            create_database(str(db.engine.url))
> > +        db.drop_all()
>
> It means that you haven't cleaned your db on tear down and your test MUST
> necessarily fail then.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inveniosoftware/invenio-accounts/pull/53/files#r47198588>
> .
>

lnielsen replied:
It's useful when you kill tests with ctrl-c during development so you don't have to clean up the DB manually in those cases.

jirikuncar replied:
IMHO it's not good idea. If you kill the tests with `ctrl-c` then you can run `db drop --yes-i-know`.

- normal runtime of tests is longer
- you might not discover that something went wrong after your tests

jirikuncar replied:
> It's useful when you kill tests with ctrl-c during development so you don't have to clean up the DB manually in those cases.

Lazy developer vs. clean code ... I'm personally against this practice, but if the majority wants to do it I'm not going to block it. I hope I gave enough good reasons so you can re-consider your decision.

lnielsen replied:
Well, it's all a matter of perspective. "Lazy developer vs. clean code" could just as well be put as "Complex usage vs. User friendly usage" depending on how you look at it ;-)

jirikuncar replied:
PyTest have you covered - it automatically calls `request.finalizers` on `KeyboardInterupt` :wink: Hence, it's even more important that you do correct clean-up. It cleans when the test fails too, hence you just need to re-run the tests if your environment is in dirty state.

otron replied:
I'm inclined to agree with @jirikuncar that it isn't necessary to drop the db during initialization, as the finalizers are run on non-exceptional interrupts and should clean out the db.

However this currently exists in the `master` branch in [`tests/conftest.py`](https://github.com/inveniosoftware/invenio-accounts/blob/master/tests/conftest.py#L78) so I'd rather keep it here to minimize the differences between the two `app` fixtures.

otron commented:
Changed this back to WIP while I look for a fix for the Travis timeouts.

otron commented:
Update on the Travis timeouts:
1. Added a timeout to the `env_browser` fixture.
 This revealed that the tests were timing out when run with PhantomJS.
 Supported by [job #331.2](https://travis-ci.org/inveniosoftware/invenio-accounts/jobs/97638304), [job #331.8](https://travis-ci.org/inveniosoftware/invenio-accounts/jobs/97638315) and [job #332.26](https://travis-ci.org/inveniosoftware/invenio-accounts/jobs/97659282)

2. Removed PhantomJS from the list of browsers in `.travis.yml`.
  While [build #333](https://travis-ci.org/inveniosoftware/invenio-accounts/builds/97663340) and [build #335](https://travis-ci.org/inveniosoftware/invenio-accounts/builds/100080238) didn't fail, this doesn't prove that the problem has been fixed as we don't have data for all previously timed out jobs.
